### ISSUE TITLE
Failure executing sqlite3 after bedde6d (at and preopen semantics)

### ISSUE THREAD

**Issue by ncruces** [tetratelabs/wazero#1027]
**Describe the bug**

After bedde6d I'm getting some rather inscrutable bugs while testing [`github.com/ncruces/go-sqlite3`](https://github.com/ncruces/go-sqlite3).

I'm not entirely sure what's going on.

**To Reproduce**

```sh
git clone --branch crash https://github.com/ncruces/go-sqlite3
cd go-sqlite3
./build_deps.sh
go run ./cmd
```

This panics inside my own code with the following error:
```
panic: module "sqlite3-1" closed with exit_code(71)

goroutine 1 [running]:
github.com/ncruces/go-sqlite3.(*Conn).free(0xc0000e26c0, 0x10cc20)
	/Users/cruces/Desktop/go-sqlite3/conn.go:166 +0xa5
github.com/ncruces/go-sqlite3.Open({0x11ab821, 0x7}, 0x0, {0x0, 0x0})
	/Users/cruces/Desktop/go-sqlite3/conn.go:92 +0xa29
main.main()
	/Users/cruces/Desktop/go-sqlite3/cmd/main.go:11 +0x2f
exit status 2
```

It's this section of code:
https://github.com/ncruces/go-sqlite3/blob/7c3b679b7dcc8fe36cf87a59447ab1048bf1b4d0/conn.go#L86-L94

After calling `c.api.open.Call` (which does a bunch of WASI FS calls) any function call seems to return this error (I'm panicking `free` errors). I'm guessing some `libc` internal state somehow gets corrupted?

**Expected behavior**

The above repro should print this:
```
2023/01/12 15:08:14 sqlite3: disk I/O error
exit status 1
````

Or, after #1026, it should succeed in creating a `test.db` with some tables in it.

**Comment by ncruces** [comment]
This seems to also break [`github.com/ncruces/rethinkraw/pkg/dcraw`](https://pkg.go.dev/github.com/ncruces/rethinkraw/pkg/dcraw).

I'll try to narrow this down. But maybe some end-to-end test to prevent this from happening in the future (something that stress tests WASI FS) would be interesting.

**Comment by ncruces** [comment]
On `dcraw` this is not an issue after all.

I had implemented a custom file system (to sandbox `dcraw` to read from a single file), and had left some stuff unimplemented that are now needed (`wazero` now opens and stats `"."` at startup).

Still need to figure out why this breaks SQLite, which will be _much_ harder, since that's _way_ too much code to track down.

**Comment by codefromthecrypt** [comment]
> I had implemented a custom file system (to sandbox dcraw to read from a single file), and had left some stuff unimplemented that are now needed (wazero now opens and stats "." at startup).

On the stat of "." at startup, it should be lazy until the wasm actually asks for a stat. We needed that to give real data back as before we were giving fake info. It may not be as critical to stat that and instead give fake info on a pre-open. If you want to change this back to fake data, it could be fine I suppose. It is mainly that the permissions and time wouldn't be valid iirc. Since you are a heavy user of WASI I think your opinion either way is valid and PR welcome if you feel it should change.

**Comment by codefromthecrypt** [comment]
PS I added the host listener which is less to track, which may be closer to identifying what's causing the panic.

```bash
$ ± |crash U:4 ?:1 ✗| → go run ./cmd
--> ._start()
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	--> .malloc(2)
	<-- 1101296
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	--> .free(0)
	<--
	--> .malloc(1)
	<-- 1101360
	--> .free(1101296)
	<--
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
	==> wasi_snapshot_preview1.random_get(buf=1048572,buf_len=4)
	<== errno=ESUCCESS
	--> .malloc(18)
	<-- 1101376
	--> .free(1101376)
	<--
<--
--> .malloc(8)
<-- 1101296
--> .malloc(4)
<-- 1101376
--> .sqlite3_open_v2(1101296,1101376,6,0)
	--> .malloc(528)
	<-- 1101392
	--> .malloc(75)
	<-- 1101936
	--> .malloc(24)
	<-- 1102032
	--> .malloc(75)
	<-- 1102064
	--> .malloc(24)
	<-- 1102160
	--> .malloc(74)
	<-- 1102192
	--> .malloc(24)
	<-- 1102288
	--> .malloc(23)
	<-- 1102320
	--> .malloc(40)
	<-- 1102352
	--> .malloc(84)
	<-- 1102400
	--> .malloc(1034)
	<-- 1102496
	--> .malloc(502)
	<-- 1103552
	--> .free(1102496)
	<--
	==> wasi_snapshot_preview1.fd_fdstat_get(fd=3)
	<== (stat={filetype=DIRECTORY,fdflags=,fs_rights_base=,fs_rights_inheriting=},errno=ESUCCESS)
	==> wasi_snapshot_preview1.path_open(fd=3,dirflags=SYMLINK_FOLLOW,path=test.db,oflags=CREAT,fs_rights_base=,fs_rights_inheriting=,fdflags=)
	<== (opened_fd=4,errno=ESUCCESS)
	--> .malloc(4112)
	<-- 1104064
	--> .malloc(128)
	<-- 1102496
	--> .malloc(1032)
	<-- 1108192
	==> wasi_snapshot_preview1.fd_seek(fd=4,offset=0,whence=0,result.newoffset=1048360)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.fd_read(fd=4,iovs=1048360,iovs_len=1)
	<== (nread=0,errno=ESUCCESS)
	--> .malloc(92)
	<-- 1102640
	--> .malloc(92)
	<-- 1102752
	--> .malloc(14)
	<-- 1102864
	--> .malloc(20)
	<-- 1102896
	--> .malloc(54)
	<-- 1102928
	--> .malloc(24)
	<-- 1102992
	--> .malloc(42)
	<-- 1103024
	--> .malloc(24)
	<-- 1103088
	--> .malloc(42)
	<-- 1103120
	--> .malloc(24)
	<-- 1103184
	--> .malloc(48008)
	<-- 1109232
	--> .free(1102320)
	<--
<-- 0
--> .free(1101376)
<--
--> .free(1101296)
<--
--> .malloc(60)
<-- 1103216
--> .sqlite3_exec(1101400,1103216,0,0,0)
	--> .malloc(14)
	<-- 1102320
	--> .malloc(22)
	<-- 1103296
	--> .malloc(72)
	<-- 1103328
	--> .malloc(13)
	<-- 1103408
	--> .malloc(20)
	<-- 1103440
	--> .malloc(13)
	<-- 1103472
	--> .free(1103440)
	<--
	--> .malloc(17)
	<-- 1103440
	--> .free(1103504)
	<--
	--> .malloc(17)
	<-- 1103504
	--> .malloc(12)
	<-- 1157312
	--> .free(1157248)
	<--
	--> .malloc(24)
	<-- 1157248
	==> wasi_snapshot_preview1.path_filestat_get(fd=3,flags=SYMLINK_FOLLOW,path=test.db-journal)
	<== (filestat=,errno=ENOENT)
	==> wasi_snapshot_preview1.path_filestat_get(fd=3,flags=SYMLINK_FOLLOW,path=test.db-wal)
	<== (filestat=,errno=ENOENT)
	==> wasi_snapshot_preview1.fd_filestat_get(fd=4)
	<== (filestat={filetype=REGULAR_FILE,size=0,mtim=1673653389866503846},errno=ESUCCESS)
	--> .malloc(85128)
2023/01/14 07:50:52 wasm error: out of bounds memory access
wasm stack trace:
	.attempt_allocate(i32,i32) i32
	.emmalloc_memalign(i32) i32
	.malloc(i32) i32
	.sqlite3MemMalloc(i32) i32
	.sqlite3Malloc(i64) i32
	.pcache1FetchStage2(i32,i32,i32) i32
	.pcache1Fetch(i32,i32,i32) i32
	.getPageNormal(i32,i32,i32,i32) i32
	.sqlite3BtreeBeginTrans(i32,i32,i32) i32
	.sqlite3InitOne(i32,i32,i32,i32) i32
	.sqlite3Init(i32,i32) i32
	.sqlite3StartTable(i32,i32,i32,i32,i32,i32,i32)
	.yy_reduce(i32,i32,i32,i32) i32
	.sqlite3RunParser(i32,i32) i32
	.sqlite3Prepare(i32,i32,i32,i32,i32,i32,i32) i32
	.sqlite3LockAndPrepare(i32,i32,i32,i32,i32,i32,i32) i32
	.sqlite3_exec(i32,i32,i32,i32,i32) i32
exit status 1
```

Here's the diff which is only updating go.mod to latest and the log listener

```diff
diff --git a/compile.go b/compile.go
index 5f6aa1e..ac0398c 100644
--- a/compile.go
+++ b/compile.go
@@ -1,7 +1,6 @@
 package sqlite3
 
 import (
-       "context"
        "os"
        "sync"
        "sync/atomic"
@@ -24,8 +23,6 @@ var (
 )
 
 func compile() {
-       ctx := context.Background()
-
        wasm = wazero.NewRuntime(ctx)
        wasi_snapshot_preview1.MustInstantiate(ctx, wasm)
 
diff --git a/conn.go b/conn.go
index b703cdd..029115a 100644
--- a/conn.go
+++ b/conn.go
@@ -3,7 +3,10 @@ package sqlite3
 import (
        "bytes"
        "context"
+       "github.com/tetratelabs/wazero/experimental"
+       "github.com/tetratelabs/wazero/experimental/logging"
        "io/fs"
+       "os"
        "path/filepath"
        "strconv"
 
@@ -12,6 +15,8 @@ import (
        "github.com/tetratelabs/wazero/experimental/writefs"
 )
 
+var ctx = context.WithValue(context.Background(), experimental.FunctionListenerFactoryKey{}, logging.NewHostLoggingListenerFactory(os.Stderr))
+
 type Conn struct {
        handle uint32
        module api.Module
@@ -32,13 +37,12 @@ func Open(name string, flags uint64, vfs string) (conn *Conn, err error) {
                }
        }
 
-       ctx := context.TODO()
-
        cfg := wazero.NewModuleConfig().
                WithName("sqlite3-" + strconv.FormatUint(counter.Add(1), 10))
        if fs != nil {
                cfg = cfg.WithFS(fs)
        }
+
        module, err := wasm.InstantiateModule(ctx, module, cfg)
        if err != nil {
                return nil, err
@@ -100,7 +104,7 @@ func Open(name string, flags uint64, vfs string) (conn *Conn, err error) {
 }
 
 func (c *Conn) Close() error {
-       r, err := c.api.close.Call(context.TODO(), uint64(c.handle))
+       r, err := c.api.close.Call(ctx, uint64(c.handle))
        if err != nil {
                return err
        }
@@ -108,13 +112,13 @@ func (c *Conn) Close() error {
        if r[0] != OK {
                return c.error(r[0])
        }
-       return c.module.Close(context.TODO())
+       return c.module.Close(ctx)
 }
 
 func (c *Conn) Exec(sql string) error {
        sqlPtr := c.newString(sql)
 
-       r, err := c.api.exec.Call(context.TODO(), uint64(c.handle), uint64(sqlPtr), 0, 0, 0)
+       r, err := c.api.exec.Call(ctx, uint64(c.handle), uint64(sqlPtr), 0, 0, 0)
        if err != nil {
                return err
        }
@@ -133,19 +137,19 @@ func (c *Conn) error(rc uint64) *Error {
        var r []uint64
 
        // string
-       r, _ = c.api.errstr.Call(context.TODO(), uint64(rc))
+       r, _ = c.api.errstr.Call(ctx, uint64(rc))
        if r != nil {
                serr.str = c.getString(uint32(r[0]), 512)
        }
 
        // message
-       r, _ = c.api.errmsg.Call(context.TODO(), uint64(c.handle))
+       r, _ = c.api.errmsg.Call(ctx, uint64(c.handle))
        if r != nil {
                serr.msg = c.getString(uint32(r[0]), 512)
        }
 
        // extended code
-       r, _ = c.api.errext.Call(context.TODO(), uint64(c.handle))
+       r, _ = c.api.errext.Call(ctx, uint64(c.handle))
        if r != nil {
                serr.ExtendedCode = int(r[0])
        }
@@ -161,14 +165,14 @@ func (c *Conn) free(ptr uint32) {
        if ptr == 0 {
                return
        }
-       _, err := c.api.free.Call(context.TODO(), uint64(ptr))
+       _, err := c.api.free.Call(ctx, uint64(ptr))
        if err != nil {
                panic(err)
        }
 }
 
 func (c *Conn) newBytes(len uint32) uint32 {
-       r, err := c.api.malloc.Call(context.TODO(), uint64(len))
+       r, err := c.api.malloc.Call(ctx, uint64(len))
        if err != nil {
                panic(err)
        }
@@ -183,7 +187,7 @@ func (c *Conn) newString(str string) uint32 {
 
        buf, ok := c.memory.Read(ptr, uint32(len(str)+1))
        if !ok {
-               c.api.free.Call(context.TODO(), uint64(ptr))
+               c.api.free.Call(ctx, uint64(ptr))
                panic("sqlite3: failed to init string")
        }
 
diff --git a/go.mod b/go.mod
index b37ac52..ab6a326 100644
--- a/go.mod
+++ b/go.mod
@@ -2,4 +2,4 @@ module github.com/ncruces/go-sqlite3
 
 go 1.19
 
-require github.com/tetratelabs/wazero v1.0.0-pre.7.0.20230111032309-2df84c679c86
+require github.com/tetratelabs/wazero v1.0.0-pre.7.0.20230113231614-59076bccf58a
diff --git a/go.sum b/go.sum
index 0eb86c0..941b966 100644
--- a/go.sum
+++ b/go.sum
@@ -1,2 +1,2 @@
-github.com/tetratelabs/wazero v1.0.0-pre.7.0.20230111032309-2df84c679c86 h1:7cZS5rLZ6SKNJrTztNy8Qg+X8vXU3D71qQSRpoTO7h4=
-github.com/tetratelabs/wazero v1.0.0-pre.7.0.20230111032309-2df84c679c86/go.mod h1:u8wrFmpdrykiFK0DFPiFm5a4+0RzsdmXYVtijBKqUVo=
+github.com/tetratelabs/wazero v1.0.0-pre.7.0.20230113231614-59076bccf58a h1:NbwgJ50Zv3EVoQSLmoEXMKy+lHPjnNsVrcagU/aT4sI=
+github.com/tetratelabs/wazero v1.0.0-pre.7.0.20230113231614-59076bccf58a/go.mod h1:u8wrFmpdrykiFK0DFPiFm5a4+0RzsdmXYVtijBKqUVo=
```

**Comment by codefromthecrypt** [comment]
FWIW the size looks wrong, but at least on my host it was zero length...

```bash
$ du test.db
0	test.db
``

**Comment by ncruces** [comment]
No, for `dcraw` I'm fine. I made the hacked filesystem that wouldn't pass [`testing/fstest`](https://pkg.go.dev/testing/fstest): that's on me. But I'm actually happy with my fix. I created something that I can reuse elsewhere.

I'm more worried with SQLite. I just open the DB, and immediately almost any call into the module exits with that error. And SQLite is a little to big to inspect.

I don't _need_ WASI here. The intent is to bypass it entirely with a custom from scratch VFS. Besides for now this is just an exploration. I just had this hitch to scratch and [Filippo](https://mastodon.social/@filippo/109393329513662703) pointed the way.

**Comment by codefromthecrypt** [comment]
lemme know if any of the above leading to the "out of bounds memory access" still smells like WASI, I'll be on and off tomorrow, but can look into it further if need be. luckily, it doesn't seem to make many calls prior to busting ;)

**Comment by ncruces** [comment]
> FWIW the size looks wrong, but at least on my host it was zero length...
> 
> ```shell
> $ du test.db
> 0	test.db
> ```

This is the expected behavior when it fails, I think. I'll try the logging. That looks really helpful.

While SQLite code is highly famed for its quality, it is also possible that [`demovfs`](https://www.sqlite.org/src/doc/trunk/src/test_demovfs.c) has a bug. It's a purposely simplified VFS that's not part of the main codebase, not as well tested, and the intent of which is to demonstrate how to build a VFS from scratch. My intent was to study that, and replace with my own, piecemeal, in Go.

Don't let this hold you back, or affect your roadmaps. I bring up these issues not because they're mission critical, but because wazero is an exciting project that I want to see flourish.

**Comment by codefromthecrypt** [comment]
If we think this is past WASI we can rename the issue to "Failure executing sqlite3" as it is maybe better marketing :) IOTW, I think you aren't alone in wanting sqlite to work, and perhaps someone else can help bring things forward!

Ack. trying to get existing codebases to work in wasm is an interesting journey and even more "interesting" when we have glitches in our host functions. I recall a lot of bumps when @jerbob92 was knocking around virtual filesystems as well. Please keep us posted and also yell if you need help regardless of wasi.

Finally, thanks for all the support, two calendar years now! In my presentation at that.us Tuesday, I already have content talking about dcraw. I think when this is working it will really help wazero (and go in general) flourish indeed as many can relate to sqlite.

**Comment by jerbob92** [comment]
@ncruces is there a reason sqlite is compiled with emmalloc? I have also seen issues where applications use things like mmap, but Emscripten doesn't support this.

**Comment by mathetake** [comment]
oh weird I think ncruces compiled sqlight with zig cc and shouldn't come with any emscripten specific thingy

**Comment by jerbob92** [comment]
The stacktrace mentions emmalloc, which is a non-default Emscripten allocator afaik.

**Comment by ncruces** [comment]
I'm building with `zig cc --target=wasm32-wasi`:
https://github.com/ncruces/go-sqlite3/blob/507e4da17b342f5f841acd344137e6f245ac4283/build_deps.sh#L16

**Comment by ncruces** [comment]
Just so this is easier to repro, I uploaded my `sqlite3.wasm` (it's not that big) to the `crash` branch. Also did the context change, and updated `go.mod` to wazero `main` (with 59076bccf58a491ab6a5df2bfb834e37602bfc73). Force pushed. 

From a clean checkout:
```sh
git clone --branch crash https://github.com/ncruces/go-sqlite3
cd go-sqlite3
go run ./cmd
```

I get a somewhat different log:
```
--> ._start.command_export()
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
<--
--> .malloc.command_export(8)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
<-- 1100736
--> .malloc.command_export(4)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
<-- 1100832
--> .sqlite3_open_v2.command_export(1100736,1100832,6,0)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EBADF)
	==> wasi_snapshot_preview1.fd_fdstat_get(fd=3)
	<== (stat={filetype=DIRECTORY,fdflags=,fs_rights_base=,fs_rights_inheriting=},errno=ESUCCESS)
	==> wasi_snapshot_preview1.path_open(fd=3,dirflags=SYMLINK_FOLLOW,path=test.db,oflags=CREAT,fs_rights_base=,fs_rights_inheriting=,fdflags=)
	<== (opened_fd=4,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_seek(fd=4,offset=0,whence=0,result.newoffset=1048360)
	<== errno=ESUCCESS
	==> wasi_snapshot_preview1.fd_read(fd=4,iovs=1048360,iovs_len=1)
	<== (nread=0,errno=ESUCCESS)
<-- 0
--> .free.command_export(1100832)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat={pr_name_len=1},errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_dir_name(fd=3)
	<== (path=/,errno=ESUCCESS)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=4)
	<== (prestat=,errno=EINVAL)
	==> wasi_snapshot_preview1.proc_exit(rval=71)
--> .sqlite3_close.command_export(1100904)
	==> wasi_snapshot_preview1.fd_prestat_get(fd=3)
	<== (prestat=,errno=EBADF)
	==> wasi_snapshot_preview1.fd_close(fd=4)
	<== errno=EBADF
<-- 0
panic: module "sqlite3-1" closed with exit_code(71)

goroutine 1 [running]:
github.com/ncruces/go-sqlite3.(*Conn).free(0xc0001ca360, 0x10cc20)
	/Users/cruces/Desktop/go-sqlite3/conn.go:166 +0xa5
github.com/ncruces/go-sqlite3.Open({0x127f3db, 0x7}, 0x0, {0x0, 0x0})
	/Users/cruces/Desktop/go-sqlite3/conn.go:92 +0xa31
main.main()
	/Users/cruces/Desktop/go-sqlite3/cmd/main.go:11 +0x2f
exit status 2
```

I find it weird that every call does a new `fd_prestat_get` for `fd`s 3 and 4? But I admit I haven't really studied these syscalls…

Then after the `sqlite3_open_v2` on the call to `free`, `fd_prestat_get` for `fd` 4 becomes `EINVAL` instead of `EBADF`, and `proc_exit` happens with exit code 71. This in turn causes the panic in my own code (I `panic` on errors in `malloc`/`free`).

**Comment by ncruces** [comment]
OK, so it seems changing this bit:
https://github.com/tetratelabs/wazero/blob/59076bccf58a491ab6a5df2bfb834e37602bfc73/imports/wasi_snapshot_preview1/fs.go#L1370-L1377

To return `ErrnoBadf` instead of `ErrnoInval` for files that are not pre-opened fixes this issue. I with that change my test succeeds and I get a `test.db` with the expected contents. But I have no idea if this is the correct fix, because, as I said above, I haven't really studied these syscalls.

**Comment by codefromthecrypt** [comment]
> It seems that, for some reason, for every call into the module, [WASI libc](https://github.com/WebAssembly/wasi-libc) is testing every fd starting from 3 (so after stdin, stdout, stderr). Probably to figure out what's the current working directory (to resolve relative paths)? I don't know. And it stops on the first EBADF. If it finds any error that's not EBADF, it kills the "process" with exit code 71 (aka. [EX_OSERR](https://github.com/WebAssembly/wasi-libc/blob/659ff414560721b1660a19685110e484a081c3d4/libc-top-half/musl/include/sysexits.h#L12))

So, what I've recall from wasi-libc is that they always assume cwd is root (zig has a different view which is first pre-open is root, but only when zig isn't using wasi-libc).

```
// We don't handle AT_FDCWD, as that's resolved in the compiler. There's no
// working directory function in WASI, so most assume CWD is "/". Notably, Zig
// has different behavior which assumes it is whatever the first pre-open name
// is.
//
// See https://github.com/WebAssembly/wasi-libc/blob/659ff414560721b1660a19685110e484a081c3d4/libc-bottom-half/sources/at_fdcwd.c
// See https://linux.die.net/man/2/openat
```

That the pre-opens logic is being called multiple times seems fishy indeed, and I would call it a perf bug really. You might look at TinyGo to see how they do it. While technically exports besides _start aren't really valid in WASI, it is routine, and I don't think TinyGo (which uses wasi-libc) re-scans the FDs per call of export. Maybe worth a look.

**Comment by ncruces** [comment]
Regarding `dcraw` I tagged [`github.com/ncruces/rethinkraw`](https://pkg.go.dev/github.com/ncruces/rethinkraw/pkg/dcraw) to ensure it works with `wazero` from head, should anyone be crazy enough to try it.

**Comment by ncruces** [comment]
On SQLite.

SQLite is among the [most widely deployed software of any kind](https://www.sqlite.org/mostdeployed.html).

At 150k lines of code, of [very high quality code](https://www.sqlite.org/testing.html), it's both highly sophisticated, but small enough that using it through WASM makes perfect sense. It's also self-contained enough to make it possible (the entire [OS interface](https://www.sqlite.org/vfs.html) is easy to pull out and reimplement in Go).

Given the community's stance on [Cgo](https://dave.cheney.net/2016/01/18/cgo-is-not-go), considerable effort [has been spent](https://pkg.go.dev/modernc.org/sqlite) to try to avoid it.

I'd been musing on the idea for a while, and [Filippo](https://mastodon.social/@filippo/109393329513662703) finally pushed me there.

**Comment by ncruces** [comment]
I don't think there's anything actionable for `wazero` to do here, but I added a [benchmark](https://github.com/ncruces/go-sqlite3/tree/main/bench) to [`github.com/ncruces/go-sqlite3`](https://github.com/ncruces/go-sqlite3) that compares it to [`crawshaw.io/sqlite`](https://pkg.go.dev/crawshaw.io/sqlite), and thought you might be curious.

The results:
```sh
$ go test -bench=. -benchmem -benchtime=10s ./bench    
goos: darwin
goarch: amd64
pkg: github.com/ncruces/go-sqlite3/bench
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkCrawshaw-12    	      18	 655783355 ns/op	104002710 B/op	 8999998 allocs/op
BenchmarkWasm-12        	      12	 995722020 ns/op	178749920 B/op	11000099 allocs/op
PASS
ok  	github.com/ncruces/go-sqlite3/bench	26.639s
```

The large memory difference I assume is due to Cgo memory not being tracked, and this being a `:memory:` database. `wazero` will have a footprint, but it should be fixed (and amortized by caching the `CompiledModule`), not "per operation." It might make sense to reuse the instantiated module (across connections), or pool connections themselves, **neither of which I'm doing**.

The 50% CPU difference is more relevant.

The _"bad"_ news there that all the `prestat` chattiness makes a big difference (and there's likely little `wazero` can do about that). The _"good"_ news for `sqlite3` is that this can mostly be avoided by implementing a [Go VFS](https://github.com/ncruces/go-sqlite3/blob/govfs/vfs.go):
```sh
$ go test -bench=. -benchmem -benchtime=10s ./bench
goos: darwin
goarch: amd64
pkg: github.com/ncruces/go-sqlite3/bench
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
BenchmarkCrawshaw-12    	      18	 663649267 ns/op	104002825 B/op	 9000000 allocs/op
BenchmarkWasm-12        	      15	 719824482 ns/op	178749060 B/op	11000105 allocs/op
PASS
ok  	github.com/ncruces/go-sqlite3/bench	25.207s
```
